tensorflow/c/kernels.h

```diff
@@ -360,6 +372,68 @@ TF_CAPI_EXPORT extern TF_Tensor* TF_AllocateTemp(
     TF_OpKernelContext* context, TF_DataType dtype, int64_t* dims, int num_dims,
     TF_AllocatorAttributes* alloc_attrs, TF_Status* status);
 
+typedef struct TF_VariableInputLockHolder TF_VariableInputLockHolder;
+
+// Expose higher level Assignment operation for Pluggable vendors to implement
+// in the plugin for Training. The API takes in the context with indices for
+// the input and value tensors. It also accepts the copy functor provided by
+// pluggable vendor to do the copying of the tensors.
+TF_CAPI_EXPORT extern void TF_AssignVariable(
+          TF_OpKernelContext* ctx,
+          int input_index,
+          int value_index,
+          void (*copyFunc)(TF_OpKernelContext * ctx,
+                           TF_Tensor *source,
+                           TF_Tensor *dest),
+          TF_Status* status);
+
+// This is a helper function which acquires mutexes in-order to provide thread-safe 
+// way of performing weights update during the optimizer op. It returns an opaque 
+// LockHolder handle back to plugin. This handle is passed to the Release API for 
+// releasing the locks when the weight update is done.
+TF_CAPI_EXPORT extern void TF_MaybeLockVariableInputMutexesInOrder(
+                                    TF_OpKernelContext* ctx,
+                                    bool do_lock, bool sparse,
+                                    const int* const input_ids,
+                                    size_t len,
+                                    void (*copyFunc)(TF_OpKernelContext * ctx,
+                                                     TF_Tensor *source,
+                                                     TF_Tensor *dest),
+                                    TF_VariableInputLockHolder** lockHolder,
+                                    TF_Status* status);
+
+// This interface returns `out` tensor which is updated corresponding to the
+// variable passed with input index.
+TF_CAPI_EXPORT extern void TF_GetInputTensorFromVariable(
+              TF_OpKernelContext* ctx, 
+              int input,
+              bool lock_held,
+              bool isVariantType,
```

Comment:
Do we need to expose `isVariantType` in the API  or can that be inferred from the tensor? I believe this is equivalent to `TF_TensorType(tensor) == TF_VARIANT`.

Reply:
@saxenasaurabh thanks! (sorry for the delay), you are right , we can probably skip the isVariantType in the API. Currently in our `metal-plugin` release we are using this API. If its not too much of a concern, can we keep this way?